File: src/abstractions/Bakabase.Abstractions/Models/Domain/Resource.cs
using Bakabase.InsideWorld.Models.Constants;
using System.Collections.Frozen;
using Bakabase.Abstractions.Extensions;
using Bakabase.Abstractions.Models.Domain.Constants;

namespace Bakabase.Abstractions.Models.Domain;

public record Resource
{
    public int Id { get; set; }

    [Obsolete]
    public int MediaLibraryId { get; set; }

    public ResourceStatus Status { get; set; } = ResourceStatus.Active;

    /// <summary>
    /// Source links for this resource. A resource can be discovered by multiple sources.
    /// </summary>
    public List<ResourceSourceLink>? SourceLinks { get; set; }

    public string? FileName => string.IsNullOrEmpty(Path) ? null : System.IO.Path.GetFileName(Path);

    public string? Directory => string.IsNullOrEmpty(Path) ? null : System.IO.Path.GetDirectoryName(Path).StandardizePath()!;

    public string? Path { get; set; }

    /// <summary>
    /// Whether this resource currently has local files. False means the resource is known
    /// to Bakabase but not materialized on disk yet — an uninstalled Steam game, a work the
    /// user intends to acquire. Prefer this over inspecting <see cref="Path"/> directly so
    /// the rule lives in one place.
    /// </summary>
    public bool HasLocalPath => !string.IsNullOrEmpty(Path);

    private string? _displayName;

    public string? DisplayName
    {
        get => _displayName ?? FileName;
        set => _displayName = value;
    }

    public int? ParentId { get; set; }
    public bool HasChildren => Tags.Contains(ResourceTag.IsParent);
    public bool IsFile { get; set; }

    public DateTime CreatedAt { get; set; }

    public DateTime UpdatedAt { get; set; }
    public DateTime FileCreatedAt { get; set; }
    public DateTime FileModifiedAt { get; set; }
    public IReadOnlySet<ResourceTag> Tags { get; set; } = FrozenSet<ResourceTag>.Empty;
    public Resource? Parent { get; set; }
    /// <summary>
    /// ResourcePropertyType - PropertyId - Property
    /// </summary>
    public Dictionary<int, Dictionary<int, Property>>? Properties { get; set; }

    /// <summary>
    /// Per-(propertyPool, propertyId) value scope preferences for this resource. Most granular layer
    /// in the resolution chain (per-resource → profile → global). Populated alongside Properties.
    /// </summary>
    public List<PropertyValueScopePreference>? ScopePreferences { get; set; }

    public bool Pinned => Tags.Contains(ResourceTag.Pinned);
    public DateTime? PlayedAt { get; set; }

    /// <summary>
    /// Aggregated health score (max-priority then min-score across enabled
    /// profiles). Populated by the resource-loading layer from
    /// <see cref="Bakabase.Abstractions.Services.IResourceHealthScoreReader"/>
    /// when the HealthScore module is wired in. Null when no profile has scored
    /// the resource.
    /// </summary>
    public decimal? HealthScore { get; set; }

    /// <summary>
    /// Final resolved cover paths for this resource, populated by service layer using priority:
    /// 1. User-set covers (ReservedProperty.Cover, scope=Manual)
    /// 2. External source local covers (SourceLink.LocalCoverPaths)
    /// 3. Enhancer covers (scope=XxxEnhancer)
    /// 4. FileSystem auto-discovered covers (from cache)
    /// </summary>
    public List<string>? Covers { get; set; }

    /// <summary>
    /// Final resolved playable items for this resource, aggregated from all sources.
    /// </summary>
    public List<PlayableItem>? PlayableItems { get; set; }

    /// <summary>
    /// Per-origin readiness states for cover, playable item, and metadata data.
    /// Populated at runtime by providers; not persisted.
    /// </summary>
    public List<ResourceDataState>? DataStates { get; set; }

    [System.Text.Json.Serialization.JsonIgnore]
    public ResourceFileSystemCache? Cache { get; set; }

    public record Property(
        string? Name,
        PropertyType Type,
        List<Property.PropertyValue>? Values,
        bool Visible = false,
        int Order = 0)
    {
        public string? Name { get; set; } = Name;
        public List<PropertyValue>? Values { get; set; } = Values;
        public PropertyType Type { get; set; } = Type;
        public StandardValueType DbValueType => PropertyTypeValueTypes.GetDbValueType(Type);
        public StandardValueType BizValueType => PropertyTypeValueTypes.GetBizValueType(Type);
        public bool Visible { get; set; } = Visible;
        public int Order { get; set; } = Order;

        public record PropertyValue(
            int Scope,
            object? Value,
            object? BizValue,
            object? AliasAppliedBizValue)
        {
            public int Scope { get; set; } = Scope;
            public object? Value { get; set; } = Value;

            /// <summary>
            /// Null means the descriptor could not interpret the stored db value — it points at a
            /// choice / tag / node that no longer exists. Falling back to <see cref="Value"/> here
            /// would leak the raw id to the UI (and, for Tags and Multilevel, a value of the wrong
            /// shape entirely, since their biz type is not ListString), defeating the miss
            /// behavior every read path is supposed to share. Callers treat null as absent.
            /// </summary>
            public object? BizValue { get; set; } = BizValue;

            /// <summary>
            /// Alias application is optional, so falling back to <see cref="BizValue"/> is correct —
            /// falling back to <see cref="Value"/> is not, for the reason above.
            /// </summary>
            public object? AliasAppliedBizValue { get; set; } = AliasAppliedBizValue ?? BizValue;

            public bool IsManuallySet => Scope == (int)PropertyValueScope.Manual;
        }
    }

    [Obsolete]
    public string? MediaLibraryName { get; set; }
    [Obsolete]
    public string? MediaLibraryColor { get; set; }

    public List<MediaLibraryInfo>? MediaLibraries { get; set; }

    public record MediaLibraryInfo(int Id, string Name, string? Color);
}

File: src/web/src/core/models/Resource.ts
import type {
  DataOrigin,
  DataStatus,
  PropertyType,
  PropertyValueScope,
  ResourceDataType,
  ResourceSource,
  ResourceStatus,
  ResourceTag,
  StandardValueType,
} from "@/sdk/constants";
import type { PropertyPool } from "@/sdk/constants";

export type Value = {
  scope: PropertyValueScope;
  value?: any;
  bizValue?: any;
  aliasAppliedBizValue?: any;
};

/**
 * Whether a scope's resolved value actually holds content. Plain truthiness is wrong here: an
 * empty array is truthy but empty, while 0 and false are falsy but valid. Used to skip empty
 * scopes when resolving a property's multi-scope values down to one displayed value.
 */
export const isNonEmptyValue = (value: any): boolean => {
  if (value == null) return false;
  if (Array.isArray(value)) return value.length > 0;
  if (typeof value === "string") return value.trim().length > 0;

  return true;
};

export type Property = {
  name?: string;
  type: PropertyType;
  dbValueType: StandardValueType;
  bizValueType: StandardValueType;
  values?: Value[];
  visible?: boolean;
  order: number;
};

export type PlayableItem = {
  origin: DataOrigin;
  key: string;
  displayName?: string;
};

export type ResourceDataState = {
  resourceId: number;
  dataType: ResourceDataType;
  origin: DataOrigin;
  status: DataStatus;
};

export type PropertyValueScopePriority = {
  scope: PropertyValueScope;
  /** When this scope is empty, whether to continue with the next scope or stop and render blank. Ignored on the last entry. */
  fallbackOnEmpty: boolean;
};

export type PropertyValueScopePreference = {
  resourceId: number;
  propertyPool: PropertyPool;
  propertyId: number;
  /** Ordered scope priorities with per-scope fallback flag; null = no override (falls through to profile/global) */
  priorities?: PropertyValueScopePriority[];
};

/**
 * Resolve the effective scope chain for one property. A per-resource preference, when it carries
 * priorities, fully replaces the global priority. Within a preference an entry with
 * `fallbackOnEmpty: false` truncates the chain there — later scopes become unreachable, so an
 * empty value at that scope renders blank instead of falling through. The global priority has no
 * flags and always falls through.
 */
export const buildEffectiveScopePriority = (
  globalPriority: PropertyValueScope[],
  preference?: PropertyValueScopePreference,
): PropertyValueScope[] => {
  if (!preference?.priorities || preference.priorities.length === 0) {
    return globalPriority;
  }

  const chain: PropertyValueScope[] = [];

  for (const p of preference.priorities) {
    chain.push(p.scope);
    if (!p.fallbackOnEmpty) break;
  }

  return chain;
};

/** Walk `effectivePriority` and return the first scope that holds a non-empty value. */
export const selectScopedValue = (
  values: Value[] | undefined,
  effectivePriority: PropertyValueScope[],
): Value | undefined => {
  for (const scope of effectivePriority) {
    const value = values?.find((v) => v.scope == scope);

    if (value && isNonEmptyValue(value.aliasAppliedBizValue ?? value.bizValue)) {
      return value;
    }
  }

  return undefined;
};

/** Resolve a property's per-scope values down to the single value to display. */
export const resolveScopedValue = (
  values: Value[] | undefined,
  globalPriority: PropertyValueScope[],
  preference?: PropertyValueScopePreference,
): Value | undefined =>
  selectScopedValue(values, buildEffectiveScopePriority(globalPriority, preference));

export type ResourceSourceLink = {
  id: number;
  resourceId: number;
  source: ResourceSource;
  sourceKey: string;
  coverUrls?: string[];
  localCoverPaths?: string[];
};

export type Resource = {
  id: number;
  mediaLibraryId: number;
  categoryId: number;
  status: ResourceStatus;
  sourceLinks?: ResourceSourceLink[];
  fileName?: string;
  directory?: string;
  displayName?: string;
  path?: string;
  /**
   * Whether the resource currently has local files. False means Bakabase knows the
   * resource but it is not materialized on disk yet — an uninstalled Steam game, a work
   * the user intends to acquire.
   */
  hasLocalPath: boolean;
  parentId?: number;
  hasChildren: boolean;
  isFile: boolean;
  createdAt: string;
  updatedAt: string;
  fileCreatedAt: string;
  fileModifiedAt: string;
  parent?: Resource;
  properties?: { [key in PropertyPool]?: Record<number, Property> };
  /** @deprecated */
  mediaLibraryName?: string;
  /** @deprecated */
  mediaLibraryColor?: string;
  mediaLibraries?: { id: number; name: string; color?: string }[];
  /** Which collections this belongs to — written-down memberships and rule matches alike. */
  collections?: { id: number; name: string; color?: string }[];
  /** @deprecated */
  category?: { id: number; name: string };
  pinned: boolean;
  tags: ResourceTag[];
  playedAt?: string;
  /** Aggregated health score; null when no profile has scored this resource. */
  healthScore?: number;
  dataStates?: ResourceDataState[];

  /** Final resolved cover paths, populated by backend using priority-based selection */
  covers?: string[];
  /** Final resolved playable items from all sources */
  playableItems?: PlayableItem[];

  /** Per-(propertyPool, propertyId) scope priority overrides for this resource */
  scopePreferences?: PropertyValueScopePreference[];

  /**
   * Client-only token, bumped when this resource is reloaded after a backend cache
   * refresh. Forces dependent UI to re-resolve from the fresh data even when a path is
   * unchanged: the cover thumbnail URL appends it to bypass the browser image cache,
   * and playable-item resolution treats a bump like a new resource so stale SSE
   * discovery results don't mask the rebuilt cache. Not sent by the backend.
   */
  reloadToken?: number;
};
